components/raftstore/src/store/fsm/apply.rs

```diff
@@ -835,33 +784,23 @@ where
                 );
             }
 
-            let res = match entry.get_entry_type() {
-                EntryType::EntryNormal => self.handle_raft_entry_normal(apply_ctx, &entry),
-                EntryType::EntryConfChange => self.handle_raft_entry_conf_change(apply_ctx, &entry),
+            if let ApplyResult::Res(res) = match entry.get_entry_type() {
+                EntryType::EntryNormal => self.handle_raft_entry_normal(apply_ctx, entry).await,
+                EntryType::EntryConfChange => {
+                    self.handle_raft_entry_conf_change(apply_ctx, entry).await
+                }
                 EntryType::EntryConfChangeV2 => unimplemented!(),
-            };
-
-            match res {
-                ApplyResult::None => {}
-                ApplyResult::Res(res) => results.push_back(res),
-                ApplyResult::Yield | ApplyResult::WaitMergeSource(_) => {
-                    // Both cancel and merge will yield current processing.
-                    apply_ctx.committed_count -= committed_entries_drainer.len() + 1;
-                    let mut pending_entries =
-                        Vec::with_capacity(committed_entries_drainer.len() + 1);
-                    // Note that current entry is skipped when yield.
-                    pending_entries.push(entry);
-                    pending_entries.extend(committed_entries_drainer);
-                    apply_ctx.finish_for(self, results);
-                    self.yield_state = Some(YieldState {
-                        pending_entries,
-                        pending_msgs: Vec::default(),
-                    });
-                    if let ApplyResult::WaitMergeSource(logs_up_to_date) = res {
-                        self.wait_merge_state = Some(WaitSourceMergeState { logs_up_to_date });
-                    }
-                    return;
+            } {
+                results.push_back(res);
+            }
+            apply_ctx.committed_count += 1;
+            if apply_ctx.kv_wb().should_write_to_engine() {
+                apply_ctx.commit(self);
```

Comment:
I find that the `commit` neglect `ApplyRes` because `results` is updated only in `finish_for`. Imagine this case:
1. a `split` admin cmd is the last entry of these committed entries.
2. After executing `split`, the `should_write_to_engine` is true and `commit` is called.
3. But the `ApplyRes` does not send to PeerFsm, this event must wait for another ApplyFsm calling `flush` when write batch is full.
I think this `commit` logic should be changed. The `results` is used for batching several cmd but if `commit` is called, `results` should push to `apply_res` and then call `flush` to persist data.

Reply:
> this event must wait for another ApplyFsm calling flush when write batch is full

If there are not incoming request, `builder.before_pause(flush_tls_ctx)` will call `flush`

Reply:
Yes. But the wait time is unnecessary.